accepted/2020/serializer/WriteableDomAndDynamic.md

```diff
@@ -0,0 +1,587 @@
+
+In the example, a JSON "NaN" string is deserialized into a `JsonValue` and when `GetValue<double>` is called, a `double` is returned, assuming the appropriate `JsonSerializerOptions` are configured for quoted numbers. When the node is serialized, it produces a JSON string (again, if configured). The consumer of the DOM is not necessarily aware that `double` values assigned to the node may be serialized to JSON as either a string or a JSON number, and after deserialization is not aware what the internal state is - it could be a `byte[]`, a `JsonElement` or a `string` depending on implementation and\or the JSON token (string or number).
+
+## Primary API
+All types live in STJ.dll.
+```cs
+namespace System.Text.Json
+// Also consider existing S.T.J.Serialization namespace instead of S.T.J.
+// since it is now coupled to the serializer rather than stand-alone.
+{
+    public abstract class JsonNode
+    {
+        internal JsonNode(); // prevent external derived classes.
+        public JsonNode(JsonSerializerOptions options = null);
+
+        public JsonSerializerOptions Options { get; }
+
+        // Support terse syntax for JsonArray (no cast necessary to JsonArray).
+        public System.Text.Json.Serialization.JsonNode? this[int index] { get; set; }
+
+        // Support terse syntax for JsonObject (no cast necessary to JsonObject).
+        public virtual System.Text.Json.Serialization.JsonNode? this[string key] { get; set; }
+
+        // Support terse syntax for JsonValue (no cast necessary to JsonValue to change a value).
+        // This represents the internal value, and no conversion is performed.
+        public virtual object Value {get; set;}
+
+        // Return the internal value or convert to T as necessary.
+        public abstract T GetValue<T>();
+        public abstract object GetValue<Type type>();
+        public abstract bool TryGetValue<T>(out T? value);
+        public abstract bool TryGetValue(Type type, out object? value);
+
+        // Convert a non-dynamic node to a dynamic.
+        public object ToDynamic();
+        // This would also work assume the linker will remove the reference to SLE if not called:
+        // public dynamic ToDynamic();
+
+        // The token type from deserialization; otherwise JsonValueKind.Unspecified.
+        public JsonValueKind ValueKind { get; }
+
+        // Serialize() wrappers which pass the value and options.
+        // These are helpers and thus can be considered an optional API feature.
+        // There are no wrappers for JsonSerializer.Deserialize* methods since the
+        // existing Deserialize* methods can be called.
+        public string Serialize();
+        public void Serialize(System.Text.Json.Utf8JsonWriter writer);
+        public System.Threading.Tasks.Task SerializeAsync(Stream utf8Json, CancellationToken cancellationToken = default);
+        public byte[] SerializeToUtf8Bytes();
```

Comment:
Given the existing JSON stack and set of APIs, the interop between the serialize and JsonNode, JsonElement, JsonDocument should be on the high-level JsonSerializer type itself. We should avoid sprinkling APIs using the `Serialize` verb on other types like this. Is there a substantial benefit of this approach?

Reply:
I agree that there we too many; originally I had no "helper" Parse\Serialize methods and there were some questions on usability of not having "Parse" methods for discoverability, so those were added. However, since then I did remove the Async and Utf8 overloads.